Implement feedback

Katerina-codes opened the pull request:
- Remove coupling of **UI** with game logic by removing all strings from the main codebase.

- Implement feedback from last PR by renaming **PlayerTypes** enum and removing enum method which returned an enum string values by using **toString()** directly on each enum. 
This PR removes all **toString()** calls and all strings are replaced by **Mark** enum.

- Implement 4x4 grid.

- Improve naming.

@MollieS @Maikon @jsuchy @felipesere

Katerina-codes committed:
Rename variable for better clarity

Katerina-codes committed:
Remove UI coupling with game logic

- grid is list of enums not list of strings

Katerina-codes committed:
Board is displayed using a list of marks

Katerina-codes committed:
Replace null with Marks EMPTY for better clarity

Katerina-codes committed:
Rename Marks enum to Mark to better clarity

Katerina-codes committed:
Correct indentation and optimise imports

Katerina-codes committed:
Simplified code

Katerina-codes committed:
Made constant private

Maikon commented on src/main/game/Mark.java:
```diff
@@ -0,0 +1,7 @@
+    X,
+    O,
+    EMPTY
+}
```
👏

felipesere replied:
Nice! Is EMPTY a legitimate `Mark` though? You can't play with it, right?
Java has new type express the difference between "here is something" and "here is nothing, go away!".
Its called `Optional<T>` ([docs](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html)).

Your grid would then become..... List<Optional<Mark>>.
It might be fun to look into it.

Maikon commented on src/main/game/CommandLineUI.java:
```diff
@@ -47,16 +47,28 @@ public String getValidMove(Board board) {
+                sb.append("\n");
+            } else if (i == 8) {
+                sb.append("\n");
+            }
```
Good job for getting this to work. Try and see if there's a pattern here that can be generalised so that you avoid using the hardcoded line boundaries (2, 5, 8).

Maikon commented on src/main/game/Computer.java:
```diff
@@ -4,23 +4,23 @@
-        List<String> possibleMoves = board.availableMoves();
+    public int move(Board board) {
+        List<Integer> possibleMoves = board.availableMoves();
         return possibleMoves.get(0);
```
I would collapse these two lines into one: `board.availableMoves().get(0)`.

Maikon commented on src/main/game/Game.java:
```diff
@@ -48,7 +48,7 @@ private boolean gameIsNotOver() {
 
     private void displayBoard() {
-        List<List<String>> activeBoard = this.board.rows();
+        List<Mark> activeBoard = this.board.grid;
```
Instead of accessing the field directly, I would suggest making the `grid` field private and provide a "getter" method in order to access it.

Maikon commented on src/main/game/HumanPlayer.java:
```diff
@@ -3,23 +3,24 @@
 
     public void playMove(Board board) {
-        String playerMove = getMove(board);
+        int playerMove = getMove(board);
```
Consistent naming. In your Computer class you called this `move` so I would drop the `get` from here too.

Maikon commented on src/main/game/Line.java:
```diff
@@ -5,17 +5,17 @@
-        for (String space : spaces) {
-            if (space.equals(mark)) {
+        for (Mark space : spaces) {
+            if (space != null && space.equals(mark)) {
```
Should this be `!= null` or `!= EMPTY`?

Maikon commented on src/test/game/BoardTest.java:
```diff
@@ -2,148 +2,147 @@
-        assertEquals(createNewLine("7", "8", "9"), rows.get(2));
+        assertEquals(createNewLine(EMPTY, EMPTY, EMPTY), rows.get(0));
+        assertEquals(createNewLine(EMPTY, EMPTY, EMPTY), rows.get(1));
+        assertEquals(createNewLine(EMPTY, EMPTY, EMPTY), rows.get(2));
```
I don't think this test is testing the correct thing but more importantly I'm not sure it adds much value. Your `rows` method is tested indirectly by other methods. Also is the `rows` method used anywhere else apart from the `Board` class?

Maikon commented on src/test/game/BoardTest.java:
```diff
@@ -2,148 +2,147 @@
     public void canReturnColumns() {
-        Board board = new Board(emptyBoard);
+        Board board = new Board();
         List<Line> rows = board.columnLines();
```
Same comment as above. The variable name is also incorrect (`columns`).

Maikon commented on src/test/game/BoardTest.java:
```diff
@@ -2,148 +2,147 @@
     }
 
     @Test
     public void playersCantEnterSameMoveWhenMoveIsO() {
```
I think this test and the above are the same so you can remove one of them. They're both testing the idea that you can make "a move" regardless of what the mark is.

Maikon reviewed (commented):
Nice cleanup 👍

Katerina-codes committed:
Remove dead code

Katerina-codes committed:
Replace magic number with constant

Katerina-codes committed:
Simplify code by removing If Statement

Katerina-codes committed:
Rename constant for better clarity

Katerina-codes committed:
Can find a winner on a line with four elements

Katerina-codes committed:
Can return rows for a four by four

Katerina-codes committed:
Can return columns for a four by four

Katerina-codes committed:
Can return diagonals for a four by four

Katerina-codes committed:
Four by four board can be displayed

Katerina-codes committed:
Each space has the correct formatting

- Previously only unmarked spaces were properly formatted with spaces.

Katerina-codes committed:
Correct how columns are created for the grid

- Scores a column win on a four by four.
- Add size to Board constuctor.
- Fix findWin() so it checks if a mark is equal to EMPTY not null.

Katerina-codes committed:
Correct indentation

Katerina-codes committed:
Make duplication more obvious

Katerina-codes committed:
Asks user to enter a choice for the board size

Katerina-codes committed:
Rename method to make it more generic

Katerina-codes committed:
Replace old Board constructor with new constructor

- New Board constructor is initialised with a size.
- Corrected logic for getting diagonals.

Katerina-codes committed:
Correct tests so they test behaviour

MollieS commented on src/main/game/gameRunner.java:
```diff
@@ -1,12 +1,18 @@
+        if (userChoice == 1) {
+            boardSize = 3;
+        } else {
+            boardSize = 4;
```
Maybe these magic numbers could be made less magic?

MollieS commented on src/test/game/BoardTest.java:
```diff
@@ -2,148 +2,223 @@
+        assertEquals(new Line(EMPTY, EMPTY, EMPTY, EMPTY), rows.get(0));
+        assertEquals(new Line(EMPTY, EMPTY, EMPTY, EMPTY), rows.get(1));
+        assertEquals(new Line(EMPTY, EMPTY, EMPTY, EMPTY), rows.get(2));
+        assertEquals(new Line(EMPTY, EMPTY, EMPTY, EMPTY), rows.get(3));
```
What is this test actually testing? Your board could be returning the same row four times and this test would still pass.

felipesere commented on src/main/game/Board.java:
```diff
@@ -3,78 +3,67 @@
+    public List<Mark> grid;
 
-    public Board(List<String> grid) {
+    public Board(int size, List<Mark> grid) {
```
Why do you need this second constructor?
Also, isn't the length of the `List<Mark>` equal to the `size` parameter?

MollieS commented on src/test/game/BoardTest.java:
```diff
@@ -2,148 +2,223 @@
+        assertTrue(board.findWin(X));
+        assertTrue(board2.findWin(X));
+        assertTrue(board3.findWin(X));
+        assertTrue(board4.findWin(X));
```
You could maybe look into parameterised tests to tidy these tests up a bit

felipesere commented on src/main/game/Board.java:
```diff
@@ -3,78 +3,67 @@
-            Line line = new Line(grid.get(i), grid.get(i + 1), grid.get(i + 2));
-            rows.add(line);
+        for (int i = 0; i < size * size; i += size) {
+            if (size == 3) {
```
Checking the size gets you through today, but what about 5x5? 6x6? It adds a lot of code.
There has to be some mathy trick! For example:
If you number each row and column individually e.g. for 3x3, you'd get rows 0, 1, 2 and columns 0, 1, 2.
One of your diagonals is ALWAYS on same pairs: (0,0), (1,1), (2,2)...

Maybe that helps?

felipesere commented on src/main/game/Board.java:
```diff
@@ -83,53 +72,91 @@ public boolean findWin(String playerMark) {
+    public Line diagonalOne() {
+        Line line = null;
+        for (int i = 0; i < size; i += 2) {
+            if (size == 3) {
```
Checking the size gets you through today, but what about 5x5? 6x6? It adds a lot of code.
There has to be some mathy trick! For example:
If you number each row and column individually e.g. for 3x3, you'd get rows 0, 1, 2 and columns 0, 1, 2.
One of your diagonals is ALWAYS on same pairs: (0,0), (1,1), (2,2)...

Maybe that helps?

felipesere commented on src/main/game/Line.java:
```diff
@@ -5,17 +5,21 @@
+    private List<Mark> spaces;
 
-    public Line(String spaceOne, String spaceTwo, String spaceThree) {
+    public Line(Mark spaceOne, Mark spaceTwo, Mark spaceThree) {
```
if change the constructor to take a list instead of three (four) individual items you can grow to AAAAANY Line you want!

felipesere commented on src/main/game/CommandLineUI.java:
```diff
@@ -47,16 +51,30 @@ public String getValidMove(Board board) {
+            }
+
+            if (grid.size() == 9) {
+                if ((i + 1) % 3 == 0) {
```
Could this be `((i + 1) % square(grid.size()) == 0)`?

Katerina-codes committed:
Make HumanPlayer convert own move within class

- Fix issue where the Computer move overrode the HumanPlayer move on
4x4.
- Change updateMove() method and test so it takes moves that don't need conversion.

Katerina-codes committed:
Tidy formatting of the board

RabeaGleissner commented on src/test/game/BoardTest.java:
```diff
@@ -2,148 +2,222 @@
+    }
+
+    @Test
+    public void canScoreWinForAColumnOnAFourByFour() {
```
I don't think you need to test every possible column win... I'm just thinking - what if your mentors ask you to create a 5x5 and a 12x12 board as well? Or even to allow the board size to be specified by the user? I think just one example of a column win board is enough for the test.
